[GStreamer][PlayReady] Accept init data as sanitized

calvaris opened the pull request:
#### 71eaa127b96be13f3b4f45b4547cd18734a502af
<pre>
[GStreamer][PlayReady] Accept init data as sanitized
<a href="https://bugs.webkit.org/show_bug.cgi?id=254159">https://bugs.webkit.org/show_bug.cgi?id=254159</a>

Reviewed by Carlos Garcia Campos.

When getting init datas from adaptive streaming technologies, it can happen that information does not come in pssh form
and we need to parse it in a custom way.

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData):
(WebCore::InitDataRegistry::sanitizeCenc):

Canonical link: <a href="https://commits.webkit.org/261914@main">https://commits.webkit.org/261914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82b5f0725b494f3f1cce5921cb8df86b0bc52db7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/5016 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121695 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117374 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13547 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6240 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/119028 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100904 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106336 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1445 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46688 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14673 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1484 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98939 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15383 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20675 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53475 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8327 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17235 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pgorszkowski-igalia commented on Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
```diff
@@ -185,12 +194,83 @@ std::optional<Vector<Ref<SharedBuffer>>> InitDataRegistry::extractKeyIDsCenc(con
+    });
+
+    xmlNode* protectionDataRootElement = xmlDocGetRootElement(protectionDataXML);
+    if (protectionDataRootElement->type != XML_ELEMENT_NODE || xmlStrcmp(protectionDataRootElement->name, reinterpret_cast<const xmlChar*>("WRMHEADER")) || !protectionDataRootElement->ns)
```
shouldn't you check also possible nullptr in case  protectionDataRootElement?

carlosgcampos commented on Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
```diff
@@ -185,12 +194,83 @@ std::optional<Vector<Ref<SharedBuffer>>> InitDataRegistry::extractKeyIDsCenc(con
+        if (!decodedKeyID)
+            return false;
+  } else
+      return false;
```
I would early return instead.

calvaris committed:
[GStreamer][PlayReady] Accept init data as sanitized
https://bugs.webkit.org/show_bug.cgi?id=254159

Reviewed by Carlos Garcia Campos.

When getting init datas from adaptive streaming technologies, it can happen that information does not come in pssh form
and we need to parse it in a custom way.

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData):
(WebCore::InitDataRegistry::sanitizeCenc):

Canonical link: https://commits.webkit.org/261914@main
